File: division_y_conquista/python/quick_sort.py
def quick_sort(arreglo):
    
    if len(arreglo) <= 1: #caso base
        return arreglo

    pivot = [arreglo.pop(0)]
    mayores = []
    menores = []

    for i in range(0, len(arreglo)):
        if arreglo[i] > pivot[0]:
            mayores.append(arreglo[i])
        else:
            menores.append(arreglo[i])

    return quick_sort(menores) + pivot + quick_sort(mayores)

def main():
    arreglo = [4,10,8,7,6,5,3,12,14,2]
    arreglo = quick_sort(arreglo)
    print(arreglo)

if __name__ == "__main__":
    main()

"""
Quick sort es un método de ordenamiento que consiste en tomar un elemento como pivot de una lista desordenada
y ubicar todos los elementos mayores que el de la lista a su derecha y los menores a la izquierda.
Así quick sort ubica cada elemento en su lugar, llamando a la misma funció tanto en la lista de la izquierda
y de la derecha recursivamente hasta que cada elemento este ordenado.

Todas las funciones recursivas necesitan un caso base que se resolverá sin llamar a la funcíon principal.
En este caso consideraremos como caso base ccuando la funcion reciba una lista de un elemento o vacía (ya que estará
ordenada por definicial inicio < fin)

Podemos necesitar un módulo para ubicar el pivote: 
"""

File: division_y_conquista/c/quick_sort.c
#include <stdio.h>

void imprimirArreglo(int arreglo[], int inicio, int final);
void swap(int arreglo[], int inicio, int final);
void quickSort(int arreglo[], int inicio, int final);

int main() {
	int arreglo[] = {4,10,8,7,6,5,3,12,14,2};
	int fin = sizeof(arreglo)/sizeof(int);

	printf("### QUICK SORT ###\n\n");
	imprimirArreglo(arreglo, 0, fin);
	printf("arreglo desordenado\n");
	quickSort(arreglo, 0, fin);
	imprimirArreglo(arreglo, 0, fin);
	printf("arreglo ordenado\n");

	return 0;
}

void quickSort(int arreglo[], int inicio, int final) {
	if ( (final - inicio) < 2 ) return;

	int pivot = inicio, ult_menor = inicio;
	int punto_medio = (inicio + final) / 2;
	swap(arreglo, punto_medio, pivot); // aqui se cambia el ELEMENTO

	for (size_t i = pivot + 1; i < final; i++) { // por que pivot +1?
		if (arreglo[i] < arreglo[pivot] ) {
			ult_menor++;
			swap(arreglo, ult_menor, i);
		}
	}

	swap(arreglo, pivot, ult_menor);
	quickSort(arreglo, inicio, ult_menor);
	quickSort(arreglo, ult_menor + 1, final);
}

void swap(int arreglo[], int pos_1, int pos_2) {
	int temp = arreglo[pos_1];
	arreglo[pos_1] = arreglo[pos_2],
	arreglo[pos_2] = temp;
}

void imprimirArreglo(int arreglo[], int inicio, int final) {
	printf("[");
	for (size_t i = 0; i < final; i++)
		printf("%d ", arreglo[i]);
	printf("] --> ");
}
